### ISSUE TITLE
Several builtin functions incompatible with seemingly compatible protocols.

### ISSUE THREAD

**Issue by mikeshardmind** [microsoft/pyright#6279]
```py
from __future__ import annotations

from collections.abc import Iterable
from typing import Protocol, TypeVar

T = TypeVar("T")


class Reduction(Protocol[T]):

    def __call__(self, *__args: T) -> T:
        ...


def element_wise_apply(func: Reduction[T], seq: Iterable[Iterable[T]]) -> Iterable[T]:
    return (*(func(*args) for args in zip(*seq, strict=True)),)


data = (1,2,3), (4,5,6), (-1, 7, 2), (4, 5, 4)


print(element_wise_apply(min, data))  # prints the expected output of (-1, 2, 2)
# Both mypy and pyright think min is incompatible with Reduction
```

<details><summary>The alternative which more readily matches a typeshed overload also fails </summary> 

```py
from __future__ import annotations

from collections.abc import Iterable
from typing import Protocol, TypeVar

T = TypeVar("T")


class Reduction(Protocol[T]):

    def __call__(self, args: Iterable[T]) -> T:
        ...


def element_wise_apply(func: Reduction[T], seq: Iterable[Iterable[T]]) -> Iterable[T]:
    return (*(func(args) for args in zip(*seq, strict=True)),)


data = (1,2,3), (4,5,6), (-1, 7, 2), (4, 5, 4)


print(element_wise_apply(min, data))
```

```
Argument of type "Overload[(__arg1: SupportsRichComparisonT@min, __arg2: SupportsRichComparisonT@min, *_args: SupportsRichComparisonT@min, key: None = None) -> SupportsRichComparisonT@min, (__arg1: _T@min, __arg2: _T@min, *_args: _T@min, key: (_T@min) -> SupportsRichComparison) -> _T@min, (__iterable: Iterable[SupportsRichComparisonT@min], *, key: None = None) -> SupportsRichComparisonT@min, (__iterable: Iterable[_T@min], *, key: (_T@min) -> SupportsRichComparison) -> _T@min, (__iterable: Iterable[SupportsRichComparisonT@min], *, key: None = None, default: _T@min) -> (SupportsRichComparisonT@min | _T@min), (__iterable: Iterable[_T1@min], *, key: (_T1@min) -> SupportsRichComparison, default: _T2@min) -> (_T1@min | _T2@min)]" cannot be assigned to parameter "func" of type "Reduction[T@element_wise_apply]" in function "element_wise_apply"
  No overloaded function matches type "(*__args: Literal[1, 2, 3, 4, 5, 6, -1, 7]) -> Literal[1, 2, 3, 4, 5, 6, -1, 7]"
```

</details>

**VS Code extension or command-line**
Via pylance, but confirmed on the playground as well

[pyre handles this correctly](https://pyre-check.org/play?input=from%20__future__%20import%20annotations%0A%0Afrom%20collections.abc%20import%20Iterable%0Afrom%20typing%20import%20Protocol%2C%20TypeVar%0A%0AT%20%3D%20TypeVar(%22T%22)%0A%0A%0Aclass%20Reduction(Protocol%5BT%5D)%3A%0A%0A%20%20%20%20def%20__call__(self%2C%20*__args%3A%20T)%20-%3E%20T%3A%0A%20%20%20%20%20%20%20%20...%0A%0A%0Adef%20element_wise_apply(func%3A%20Reduction%5BT%5D%2C%20seq%3A%20Iterable%5BIterable%5BT%5D%5D)%20-%3E%20Iterable%5BT%5D%3A%0A%20%20%20%20return%20(*(func(*args)%20for%20args%20in%20zip(*seq%2C%20strict%3DTrue))%2C)%0A%0A%0Adata%20%3D%20(1%2C2%2C3)%2C%20(4%2C5%2C6)%2C%20(-1%2C%207%2C%202)%2C%20(4%2C%205%2C%204)%0A%0A%0Areveal_type(element_wise_apply(min%2C%20data))%20%0A)

Note that this is intended to work with more than just min/max or I could have just written something less generic. It's meant to be slightly more readable than inlining that everywhere.

The issue seems to persist across several builtin functions and much of `operator` and `functools` where a compatible protocol which matches *a* use of something in the standard library isn't allowed.

**Comment by erictraut** [comment]
I think pyright is correct here. The type definition of `min` requires that at least two positional arguments are passed, but the protocol supports zero or one positional arguments. That means `min` doesn't match the `Reduction` protocol as it's defined.

If you change it to the following, it passes type checking in both pyright and mypy. If pyre allows this to pass without an error, you might want to file a bug.

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChHRIoBjMAGy7XZrr0AdMQBG7bHgJEAkkjQgxvFuGhIAnriykA5pPyEoABXBIwnLgBooAFU1oAasRDMbUALy37TkAAoARDb%2BAJTMjOxcxPT0UABKaAAmCPy0pL4mlObcANo2ALrBAFyMUKVQCWjAsDDsxDxwvvRoXMDWcM46AIyFtm0wHQBMPTbWAFTtIDr0w8FQALQAfLbFZatQQhthFVXNaBBopEgwAO5YTf24uFzqvoik7D3xSSl0uXnWTQCOPXIKSmjZX6KUS8N4FeZLIH-N4rMroFAgUhQXyjW4Ie4ojr0WbAAgkSYxbRQABeWFwKK%2BHyQICw-HcNhACDQwWCllCzAS1GIHmRnWsA2sAGZWciACzWACs1gAbCLfHM%2BVAAOz8uXiqBSqCi9mMXA0w6%2BXb7Q4nM4YYiXa6%2BCDaaycqgsxhAA)

```python
class Reduction(Protocol[T]):
    def __call__(self, __arg1: T, __arg2: T, *__args: T) -> T:
        ...
```

**Comment by mikeshardmind** [comment]
@Erictraut some of the overloads in the type shed should be compatible with the collapsed alternative. not all of  the overloads require 2 positional arguments

Additionally, that protocol would change the intent.

**Comment by erictraut** [comment]
Here's the definition of `min`:
```python
@overload
def min(
    __arg1: SupportsRichComparisonT, __arg2: SupportsRichComparisonT, *_args: SupportsRichComparisonT, key: None = None
) -> SupportsRichComparisonT: ...
@overload
def min(__arg1: _T, __arg2: _T, *_args: _T, key: Callable[[_T], SupportsRichComparison]) -> _T: ...
@overload
def min(__iterable: Iterable[SupportsRichComparisonT], *, key: None = None) -> SupportsRichComparisonT: ...
@overload
def min(__iterable: Iterable[_T], *, key: Callable[[_T], SupportsRichComparison]) -> _T: ...
@overload
def min(__iterable: Iterable[SupportsRichComparisonT], *, key: None = None, default: _T) -> SupportsRichComparisonT | _T: ...
@overload
def min(__iterable: Iterable[_T1], *, key: Callable[[_T1], SupportsRichComparison], default: _T2) -> _T1 | _T2: ...
```

Which overload do you think would match the protocol in your original code sample?

**Comment by mikeshardmind** [comment]
In the collapsed section that the summary explicitly says more readily matches an overload

```py
@overload
def min(__iterable: Iterable[_T], *, key: Callable[[_T], SupportsRichComparison]) -> _T: ...
```

This overload should be matched, this is the difference contained there, both pyright and mypy fail this, pyre handles it correctly.

```py
class Reduction(Protocol[T]):

    def __call__(self, args: Iterable[T]) -> T:
        ...

...
     return (*(func(args) for args in zip(*seq, strict=True)),)
```

**Comment by erictraut** [comment]
The code in the collapsed section violates the protocol because it requires that the callable accept a keyword argument named `args`, but the parameters for `min` are positional only. If you change your protocol to be positional-only, the type error will go away.

```python
class Reduction(Protocol[T]):
    def __call__(self, args: Iterable[T], /) -> T:
        ...
```

**Comment by mikeshardmind** [comment]
Okay, I missed that while being frustrated with reproducing this and the absolutely inscrutable errors on it, and this is something I should have known, I appreciate pointing that out and I can use that here.

I think pyre is making the correct decision with the first case, but there's sort of an assumption that is inexpressible in the type system currently that I'm not wasting time using something meant to handle multidimensional data on data which isn't yet has the type of multidimensional data.

**Comment by erictraut** [comment]
I agree that the error messages here are not as useful as they could be. It's very difficult to present useful errors when overloads are involved and none of the overloads match. Pyright has some heuristics in some code paths where it attempts to "guess" what overload you might have intended and then present more detailed errors based on that best-guess overload signature, but it doesn't do so in the case where you're passing a callable to higher-order function like in your example.
